File: address-geocode-international-dot-net/REST/AGIResponse.cs
using System.Collections.Generic;
using System.Linq;
using System.Runtime.Serialization;

namespace address_geocode_international_dot_net.REST
{
    /// <summary>
    /// Represents the AGI ReverseSearch API response (REST).
    /// </summary>
    [DataContract]
    public class AGIReverseSearchResponse
    {
        public AGIReverseSearchResponse()
        {
            Locations = new List<Location>();
        }

        [DataMember(Name = "SearchInfo")]
        public SearchInfo? SearchInfo { get; set; }

        [DataMember(Name = "Locations")]
        public List<Location> Locations { get; set; }

        [DataMember(Name = "Error")]
        public ErrorDetails? Error { get; set; }

        public override string ToString()
        {
            string output = "AGI Reverse Search Response:\n";
            output += SearchInfo != null ? SearchInfo.ToString() : "  SearchInfo: null\n";

            output += Locations != null && Locations.Count > 0
                ? "  Locations:\n" +
                    string.Join("", Locations.Select((loc, i) => $"    [{i + 1}]:\n{loc}"))
                : "  Locations: []\n";

            output += Error != null ? Error.ToString() : "  Error: null\n";
            return output;
        }
    }

    /// <summary>
    /// Represents the AGI PlaceSearch API response (REST).
    /// </summary>
    [DataContract]
    public class AGIPlaceSearchResponse
    {
        [DataMember(Name = "SearchInfo")]
        public SearchInfo? SearchInfo { get; set; }

        [DataMember(Name = "Locations")]
        public Location[]? Locations { get; set; }

        [DataMember(Name = "Error")]
        public ErrorDetails? Error { get; set; }

        public override string ToString()
        {
            string output = "AGI Place Search Response:\n";
            output += SearchInfo != null ? SearchInfo.ToString() : "  SearchInfo: null\n";

            if (Locations != null && Locations.Length > 0)
            {
                output += "  Locations:\n";
                for (int i = 0; i < Locations.Length; i++)
                {
                    output += $"    [{i + 1}]:\n{Locations[i]}\n";
                }
            }
            else
            {
                output += "  Locations: []\n";
            }

            output += Error != null ? Error.ToString() : "  Error: null\n";
            return output;
        }
    }

    /// <summary>
    /// Search information block returned by the AGI service.
    /// </summary>
    [DataContract]
    public class SearchInfo
    {
        [DataMember(Name = "Status")]
        public string? Status { get; set; }

        [DataMember(Name = "NumberOfLocations")]
        public int? NumberOfLocations { get; set; }

        [DataMember(Name = "Notes")]
        public string? Notes { get; set; }

        [DataMember(Name = "NotesDesc")]
        public string? NotesDesc { get; set; }

        [DataMember(Name = "Warnings")]
        public string? Warnings { get; set; }

        [DataMember(Name = "WarningDesc")]
        public string? WarningDesc { get; set; }

        public override string ToString()
        {
            string output = "  SearchInfo:\n";
            output += $"    Status           : {Status}\n";
            output += $"    NumberOfLocations: {NumberOfLocations}\n";
            output += $"    Notes            : {Notes}\n";
            output += $"    NotesDesc        : {NotesDesc}\n";
            output += $"    Warnings         : {Warnings}\n";
            output += $"    WarningDesc      : {WarningDesc}\n";
            return output;
        }
    }

    /// <summary>
    /// Represents a geocoded location as returned by the AGI service.
    /// </summary>
    [DataContract]
    public class Location
    {
        public Location()
        {
            AddressComponents = new AddressComponents();
        }

        [DataMember(Name = "PrecisionLevel")]
        public double PrecisionLevel { get; set; }

        [DataMember(Name = "Type")]
        public string? Type { get; set; }

        [DataMember(Name = "Latitude")]
        public string? Latitude { get; set; }

        [DataMember(Name = "Longitude")]
        public string? Longitude { get; set; }

        [DataMember(Name = "AddressComponents")]
        public AddressComponents? AddressComponents { get; set; }

        [DataMember(Name = "PlaceName")]
        public string? PlaceName { get; set; }

        [DataMember(Name = "GoogleMapsURL")]
        public string? GoogleMapsURL { get; set; }

        [DataMember(Name = "BingMapsURL")]
        public string? BingMapsURL { get; set; }

        [DataMember(Name = "MapQuestURL")]
        public string? MapQuestURL { get; set; }

        [DataMember(Name = "StateFIPS")]
        public string? StateFIPS { get; set; }

        [DataMember(Name = "CountyFIPS")]
        public string? CountyFIPS { get; set; }

        [DataMember(Name = "ClassFP")]
        public string? ClassFP { get; set; }

        public override string ToString()
        {
            string output = "";
            output += $"      Type          : {Type}\n";
            output += $"      PrecisionLevel: {PrecisionLevel}\n";
            output += $"      Latitude      : {Latitude}\n";
            output += $"      Longitude     : {Longitude}\n";
            output += $"      PlaceName     : {PlaceName}\n";
            output += $"      GoogleMapsURL : {GoogleMapsURL}\n";
            output += $"      BingMapsURL   : {BingMapsURL}\n";
            output += $"      MapQuestURL   : {MapQuestURL}\n";
            output += $"      StateFIPS     : {StateFIPS}\n";
            output += $"      CountyFIPS    : {CountyFIPS}\n";
            output += $"      ClassFP       : {ClassFP}\n";
            output += AddressComponents != null
                ? AddressComponents.ToString()
                : "      AddressComponents: null\n";
            return output;
        }
    }

    /// <summary>
    /// Address components returned within a location result.
    /// </summary>
    [DataContract]
    public class AddressComponents
    {
        [DataMember(Name = "PremiseNumber")]
        public string? PremiseNumber { get; set; }

        [DataMember(Name = "Thoroughfare")]
        public string? Thoroughfare { get; set; }

        [DataMember(Name = "DoubleDependentLocality")]
        public string? DoubleDependentLocality { get; set; }

        [DataMember(Name = "DependentLocality")]
        public string? DependentLocality { get; set; }

        [DataMember(Name = "Locality")]
        public string? Locality { get; set; }

        [DataMember(Name = "AdministrativeArea1")]
        public string? AdministrativeArea1 { get; set; }

        [DataMember(Name = "AdministrativeArea1Abbreviation")]
        public string? AdministrativeArea1Abbreviation { get; set; }

        [DataMember(Name = "AdministrativeArea2")]
        public string? AdministrativeArea2 { get; set; }

        [DataMember(Name = "AdministrativeArea2Abbreviation")]
        public string? AdministrativeArea2Abbreviation { get; set; }

        [DataMember(Name = "AdministrativeArea3")]
        public string? AdministrativeArea3 { get; set; }

        [DataMember(Name = "AdministrativeArea3Abbreviation")]
        public string? AdministrativeArea3Abbreviation { get; set; }

        [DataMember(Name = "AdministrativeArea4")]
        public string? AdministrativeArea4 { get; set; }

        [DataMember(Name = "AdministrativeArea4Abbreviation")]
        public string? AdministrativeArea4Abbreviation { get; set; }

        [DataMember(Name = "PostalCode")]
        public string? PostalCode { get; set; }

        [DataMember(Name = "Country")]
        public string? Country { get; set; }

        [DataMember(Name = "CountryISO2")]
        public string? CountryISO2 { get; set; }

        [DataMember(Name = "CountryISO3")]
        public string? CountryISO3 { get; set; }

        [DataMember(Name = "GoogleMapsURL")]
        public string? GoogleMapsURL { get; set; }

        [DataMember(Name = "PlaceName")]
        public string? PlaceName { get; set; }

        [DataMember(Name = "IsUnincorporated")]
        public string? IsUnincorporated { get; set; }

        [DataMember(Name = "TimeZone_UTC")]
        public string? TimeZone_UTC { get; set; }

        [DataMember(Name = "CongressCode")]
        public string? CongressCode { get; set; }

        [DataMember(Name = "CensusTract ")]
        public string? CensusTract { get; set; }

        [DataMember(Name = "CensusGeoID ")]
        public string? CensusGeoID { get; set; }

        [DataMember(Name = "ClassFP")]
        public string? ClassFP { get; set; }

        [DataMember(Name = "StateFIPS ")]
        public string? StateFIPS { get; set; }

        [DataMember(Name = "SLDUST ")]
        public string? SLDUST { get; set; }

        [DataMember(Name = "SLDLST ")]
        public string? SLDLST { get; set; }

        [DataMember(Name = "CountyFIPS ")]
        public string? CountyFIPS { get; set; }

        [DataMember(Name = "CensusBlock ")]
        public string? CensusBlock { get; set; }
        public override string ToString()
        {
            string output = "      AddressComponents:\n";
            output += $"        PremiseNumber                  : {PremiseNumber}\n";
            output += $"        Thoroughfare                   : {Thoroughfare}\n";
            output += $"        DoubleDependentLocality        : {DoubleDependentLocality}\n";
            output += $"        DependentLocality              : {DependentLocality}\n";
            output += $"        Locality                       : {Locality}\n";
            output += $"        AdministrativeArea1            : {AdministrativeArea1}\n";
            output += $"        AdministrativeArea1Abbreviation: {AdministrativeArea1Abbreviation}\n";
            output += $"        AdministrativeArea2            : {AdministrativeArea2}\n";
            output += $"        AdministrativeArea2Abbreviation:{AdministrativeArea2Abbreviation}\n";
            output += $"        AdministrativeArea3            :{AdministrativeArea3}\n";
            output += $"        AdministrativeArea3Abbreviation:{AdministrativeArea3Abbreviation}\n";
            output += $"        AdministrativeArea4            :{AdministrativeArea4}\n";
            output += $"        AdministrativeArea4Abbreviation:{AdministrativeArea4Abbreviation}\n";
            output += $"        PostalCode                     : {PostalCode}\n";
            output += $"        Country                        : {Country}\n";
            output += $"        CountryISO2                    : {CountryISO2}\n";
            output += $"        CountryISO3                    : {CountryISO3}\n";
            output += $"        TimeZone_UTC                   : {TimeZone_UTC}\n";
            output += $"        CongressCode                   : {CongressCode}\n";
            output += $"        GoogleMapsURL                  : {GoogleMapsURL}\n";
            output += $"        PlaceName                      : {PlaceName}\n";
            output += $"        IsUnincorporated               : {IsUnincorporated}\n";
            output += $"        StateFIPS                      : {StateFIPS}\n";
            output += $"        CountyFIPS                     : {CountyFIPS}\n";
            output += $"        CensusTract                    : {CensusTract}\n";
            output += $"        CensusBlock                    : {CensusBlock}\n";
            output += $"        CensusGeoID                    : {CensusGeoID}\n";
            output += $"        ClassFP                        : {ClassFP}\n";
            output += $"        SLDUST                         : {SLDUST}\n";
            output += $"        SLDLST                         : {SLDLST}\n";


            return output;
        }
    }

    /// <summary>
    /// Error block returned if the API fails or receives invalid input.
    /// </summary>
    [DataContract]
    public class ErrorDetails
    {
        [DataMember(Name = "Type")]
        public string? Type { get; set; }

        [DataMember(Name = "TypeCode")]
        public string? TypeCode { get; set; }

        [DataMember(Name = "Desc")]
        public string? Desc { get; set; }

        [DataMember(Name = "DescCode")]
        public string? DescCode { get; set; }

        public override string ToString()
        {
            string output = "  Error:\n";
            output += $"    Type    : {Type}\n";
            output += $"    TypeCode: {TypeCode}\n";
            output += $"    Desc    : {Desc}\n";
            output += $"    DescCode: {DescCode}\n";
            return output;
        }
    }
}

File: address-geocode-international-nodejs/REST/agi_response.js
export class SearchInfo {
    constructor(data = {}) {
        this.Status = data.Status;
        this.NumberOfLocations = data.NumberOfLocations || 0;
        this.Notes = data.Notes;
        this.NotesDesc = data.NotesDesc;
        this.Warnings = data.Warnings;
        this.WarningDesc = data.WarningDesc;
    }

    toString() {
        return `Status: ${this.Status}, ` +
            `NumberOfLocations: ${this.NumberOfLocations}, ` +
            `Notes: ${this.Notes}, ` +
            `NotesDesc: ${this.NotesDesc}, ` +
            `Warnings: ${this.Warnings}, ` +
            `WarningDesc: ${this.WarningDesc}`;
    }
}

export class AddressComponents {
    constructor(data = {}) {
        this.PremiseNumber = data.PremiseNumber;
        this.Thoroughfare = data.Thoroughfare;
        this.DoubleDependentLocality = data.DoubleDependentLocality;
        this.DependentLocality = data.DependentLocality;
        this.Locality = data.Locality;
        this.AdministrativeArea1 = data.AdministrativeArea1;
        this.AdministrativeArea1Abbreviation = data.AdministrativeArea1Abbreviation;
        this.AdministrativeArea2 = data.AdministrativeArea2;
        this.AdministrativeArea2Abbreviation = data.AdministrativeArea2Abbreviation;
        this.AdministrativeArea3 = data.AdministrativeArea3;
        this.AdministrativeArea3Abbreviation = data.AdministrativeArea3Abbreviation;
        this.AdministrativeArea4 = data.AdministrativeArea4;
        this.AdministrativeArea4Abbreviation = data.AdministrativeArea4Abbreviation;
        this.PostalCode = data.PostalCode;
        this.Country = data.Country;
        this.CountryISO2 = data.CountryISO2;
        this.CountryISO3 = data.CountryISO3;
        this.GoogleMapsURL = data.GoogleMapsURL;
        this.PlaceName = data.PlaceName;
        this.IsUnincorporated = data.IsUnincorporated;
        this.StateFIPS = data.StateFIPS;
        this.CountyFIPS = data.CountyFIPS;
        this.CensusTract = data.CensusTract;
        this.CensusBlock = data.CensusBlock;
        this.CensusGeoID = data.CensusGeoID;
        this.ClassFP = data.ClassFP;
        this.CongressCode = data.CongressCode;
        this.SLDUST = data.SLDUST;
        this.SLDLST = data.SLDLST;
        this.Timezone_UTC = data.Timezone_UTC;
    }

    toString() {
        return `PremiseNumber: ${this.PremiseNumber}, ` +
            `Thoroughfare: ${this.Thoroughfare}, ` +
            `DoubleDependentLocality: ${this.DoubleDependentLocality}, ` +
            `DependentLocality: ${this.DependentLocality}, ` +
            `Locality: ${this.Locality}, ` +
            `AdministrativeArea1: ${this.AdministrativeArea1}, ` +
            `AdministrativeArea1Abbreviation: ${this.AdministrativeArea1Abbreviation}, ` +
            `AdministrativeArea2: ${this.AdministrativeArea2}, ` +
            `AdministrativeArea2Abbreviation: ${this.AdministrativeArea2Abbreviation}, ` +
            `AdministrativeArea3: ${this.AdministrativeArea3}, ` +
            `AdministrativeArea3Abbreviation: ${this.AdministrativeArea3Abbreviation}, ` +
            `AdministrativeArea4: ${this.AdministrativeArea4}, ` +
            `AdministrativeArea4Abbreviation: ${this.AdministrativeArea4Abbreviation}, ` +
            `PostalCode: ${this.PostalCode}, ` +
            `Country: ${this.Country}, ` +
            `CountryISO2: ${this.CountryISO2}, ` +
            `CountryISO3: ${this.CountryISO3}, ` +
            `GoogleMapsURL: ${this.GoogleMapsURL}, ` +
            `PlaceName: ${this.PlaceName}, ` +
            `IsUnincorporated: ${this.IsUnincorporated}, ` +
            `StateFIPS: ${this.StateFIPS}, ` +
            `CountyFIPS: ${this.CountyFIPS}, ` +
            `CensusTract: ${this.CensusTract}, ` +
            `CensusBlock: ${this.CensusBlock}, ` +
            `CensusGeoID: ${this.CensusGeoID}, ` +
            `ClassFP: ${this.ClassFP}, ` +
            `CongressCode: ${this.CongressCode}, ` +
            `SLDUST: ${this.SLDUST}, ` +
            `SLDLST: ${this.SLDLST}, ` +
            `Timezone_UTC: ${this.Timezone_UTC}`;
    }
}
export class LocationInfo {
    constructor(data = {}) {
        this.PrecisionLevel = data.PrecisionLevel || 0;
        this.Type = data.Type;
        this.Latitude = data.Latitude;
        this.Longitude = data.Longitude;
        this.AddressComponents = data.AddressComponents ? new AddressComponents(data.AddressComponents) : null;
    }

    toString() {
        return `PrecisionLevel: ${this.PrecisionLevel}, ` +
            `Type: ${this.Type}, ` +
            `Latitude: ${this.Latitude}, ` +
            `Longitude: ${this.Longitude}, ` +
            `AddressComponents: ${this.AddressComponents ? this.AddressComponents.toString() : 'null'}`;
    }
}
export class Error {
    constructor(data = {}) {
        this.Type = data.Type;
        this.TypeCode = data.TypeCode;
        this.Desc = data.Desc;
        this.DescCode = data.DescCode;
    }

    toString() {
        return `Type: ${this.Type}, TypeCode: ${this.TypeCode}, Desc: ${this.Desc}, DescCode: ${this.DescCode}`;
    }
}

export class PSResponse {
    constructor(data = {}) {
        this.SearchInfo = data.SearchInfo ? new SearchInfo(data.SearchInfo) : null;
        this.Locations = (data.Locations || []).map(location => new LocationInfo(location));
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        const locationsString = this.Locations.length
            ? this.Locations.map(location => location.toString()).join(', ')
            : 'None';
        return `PSResponse: SearchInfo = ${this.SearchInfo ? this.SearchInfo.toString() : 'null'}, ` +
            `Locations = [${locationsString}], ` +
            `Error = ${this.Error ? this.Error.toString() : 'null'}`;
    }
}
export class RSResponse {
    constructor(data = {}) {
        this.SearchInfo = data.SearchInfo ? new SearchInfo(data.SearchInfo) : null;
        this.Locations = (data.Locations || []).map(location => new LocationInfo(location));
        this.Error = data.Error ? new Error(data.Error) : null;
    }

    toString() {
        let result = 'RSResponse:\n';
        if (this.SearchInfo) {
            result += `SearchInfo: ${this.SearchInfo.toString()}\n`;
        }
        if (this.Locations && this.Locations.length > 0) {
            result += 'Locations:\n';
            result += this.Locations.map(location => location.toString()).join('\n') + '\n';
        }
        if (this.Error) {
            result += `Error: ${this.Error.toString()}`;
        }
        return result;
    }
}

export default { PSResponse, RSResponse };